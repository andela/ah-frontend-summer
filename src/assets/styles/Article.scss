.ArticleBody {
    padding: 30px;
    font-size: 1.2rem;
}

.SubmainContainer {
    display: grid;
    grid-template-columns: 1fr 1fr;
    padding: 20px;
    align-content: center;
    justify-content: center
}

.actionButtons {
    padding: 10px;
}

.LikeDislikeButton {
    border: none;
    cursor: pointer;
    transition: .3s all;
    margin-left: 15px;
    outline: none;
    background: transparent;
}

.LikeDislikeButton:active{
    transform: translateY(2px);
}
