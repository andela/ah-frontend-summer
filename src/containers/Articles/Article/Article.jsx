import React, { Component } from "react";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom"
import { toast } from "react-toastify";

import * as actions from "../../../store/actions/index";
import ArticleDetails from "../../../components/UI/Article/Article";
import DeleteArticleButtons from "../../../components/UI/Article/DeleteArticle";
import Loader from "../../../components/UI/Loader";
import {dangerToast, infoToast } from "../../../components/Toast/toast";
import Modal from "../../../components/UI/Modal/Modal";

export class Article extends Component {

    state = {
        showModal: false
    }

    componentWillMount(){
        this.props.onGetArticle(this.props.match.params.slug);
    }

    deleteArticleHandeler = () => {
        toast.dismiss();
        this.props.onDeleteArticle(this.props.match.params.slug);
    }

    showModalHandeler = () => {
        this.setState({ showModal: true });
    }

    closeModalHandler = () => {
        this.setState({ showModal: false });
    }

    render() {
        this.renderArticle = ( <Loader /> );
        this.created_at = null;
        this.isAuthor = false;
        this.isAuthenticated = localStorage.getItem("token") !== null;
        this.username = localStorage.getItem("username");
        const { article, loading, error } = this.props;
        this.confirmButtons = null;
        if (this.state.showModal) {
            this.confirmButtons = (
                <DeleteArticleButtons
                    dismissModal={this.closeModalHandler}
                    deleteArticle={this.deleteArticleHandeler}/>
            );
        };

        if ( article && article.article === "Article has been deleted") {
            infoToast(article.article);
            this.renderArticle = (<Redirect to="/"/>);
        } else if (article) {
            if (this.username) {
                this.isAuthor = ( this.username ===
                    article.author.username
                    && this.isAuthenticated)
            };
            this.created_at = article.created_at.substr(0,
                article.created_at.indexOf("T"));
            this.renderArticle = (
                <ArticleDetails
                    title={article.title}
                    author={article.author.username}
                    body={article.body}
                    readTime={article.read_time}
                    deleteArticle={this.showModalHandeler}
                    slug={this.props.match.params.slug}
                    loading={loading}
                    dateCreated={this.created_at}
                    isAuthor={this.isAuthor}
                    image={article.image}
                    />);
        } else if ( error ) {
            dangerToast(error.articles.errors);
            this.renderArticle = (<Redirect to="/"/>);
        };

        return (
            <div>
                <Modal
                show={this.state.showModal}
                modalClosed={this.closeModalHandler}>{this.confirmButtons}</Modal>
                {this.renderArticle}
            </div>
        );
    }
}

export const mapStateToProps = (state) => {
    return {
        article: state.article.article,
        loading: state.article.loading,
        error: state.article.error
    };
};

export const mapDispatchToProps = (dispatch) => {
    return {
        onGetArticle: (slug) => dispatch(actions.getArticle(slug)),
        onDeleteArticle: (slug) => dispatch(actions.deleteArticle(slug))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Article);
