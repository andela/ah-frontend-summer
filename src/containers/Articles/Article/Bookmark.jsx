import React, {Component} from 'react';
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {withRouter} from "react-router-dom";
import Bookmark from "../../../components/article/Bookmark";
import {
    bookmarkAction,
    undoBookmarkAction
} from "../../../store/actions/async/ArticleActions";
import {isLoggedIn} from "../../../store/actions/async";

export class BookmarkContainer extends Component{
    bookmarkArticle = () => {
        const {history} = this.props;
        const loggedIn = isLoggedIn(localStorage);
        if (!loggedIn){
            history.push('/login');
        }
        else{
            this.triggerBookmark();
        }
    };

    triggerBookmark = () => {
        const {bookmark, undoBookmark, isBookmarked, slug, loading} = this.props;

        if (!loading){
            if (isBookmarked){
                undoBookmark(slug);
            }
            else {
                bookmark(slug);
            }
        }
    };

    render() {
        const {loading, isBookmarked} = this.props;
        return (
            <Bookmark
                loading={loading}
                bookmark={() => {this.bookmarkArticle();}}
                isBookmarked={isBookmarked} />
        );
    }
}

export const mapStateToProps = (state) => {
    const {loading, status} = state.progress.bookmark;
    const {favorited: isBookmarked} = state.article.article;
    return {loading, status, isBookmarked: isBookmarked};
};

export const mapDispatchToProps = (dispatch) => {
    return {
        bookmark: (slug) => {
            dispatch(bookmarkAction(slug));
        },
        undoBookmark: (slug) => {
            dispatch(undoBookmarkAction(slug));
        }
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BookmarkContainer));

BookmarkContainer.propTypes = {
    bookmark: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    isBookmarked: PropTypes.bool.isRequired,
    loading: PropTypes.bool.isRequired,
    slug: PropTypes.string.isRequired,
    undoBookmark: PropTypes.func.isRequired
};
