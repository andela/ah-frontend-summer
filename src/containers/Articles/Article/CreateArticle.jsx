import React, { Component } from "react";
import { connect } from "react-redux";
import { EditorState } from "draft-js";
import { Redirect } from "react-router-dom";
import PropTypes from "prop-types";

import * as actions from "../../../store/actions/index";
import CreateArticleForm from "../../../components/UI/Article/ArticleForm";
import { successToast } from "../../../components/Toast/toast";
import sampleTags from '../../../assets/utils/sampleTags';

const suggestions = sampleTags.map((tag) => {
    return {
        id: tag,
        text: tag
    };
});
export class CreateArticle extends Component {

    state = {
        editorState: EditorState.createEmpty(),
        articleData: {
            title: "",
            description: "",
            body: "",
            image: null
        },
        error: false,
        titleTouched: false,
        descriptionTouched: false,
        bodyTouched: false,
        disabled: false,
        suggestions: suggestions,
        tags: []
    };

    onChangeHandler = (editorState) => {
        this.setState({editorState});
    };

    bodyChangedHandler=( event, editor ) => {
        const data = editor.getData();
        const articleData = {...this.state.articleData};
        articleData.body = data;
        const { body, title, description } = articleData;
        let disableBtn = true;
        if ( !body || body.length < 5 || !title || !description) {
            disableBtn = false;
        }

        this.setState({
            articleData,
            disabled: disableBtn
        });
    }

    inputHandler = (e) => {
        const articleData = {...this.state.articleData};
        let titleTouched = false;
        let descriptionTouched = false;
        let bodyTouched = false;
        let disableBtn = true;
        const { name, value, files } = e.target;

        if (name === "title") {
            articleData.title = value;
            titleTouched = false;
        } else if (name === "description") {
            articleData.description = value;
            descriptionTouched = false;
        }
        const { body, title, description } = articleData;

        if (title) {
            disableBtn = title.length > 3 && disableBtn;
        }
        if (description) {
            disableBtn = description.length > 3 && disableBtn;
        }
        if ( !title || !description || !body) {
            disableBtn = false;
        }
        if (files) {
            articleData.image = files[0];
        }

        this.setState ({
            articleData,
            titleTouched,
            descriptionTouched,
            bodyTouched,
            disabled: disableBtn
        });
        e.preventDefault();
    }

    createArticleSubmitHandler = () => {
        let disableBtnAgain = false;
        this.setState({ disabled: disableBtnAgain });
        const {
            image,
            title,
            body,
            description
        } = this.state.articleData;
        const { tags } = this.state;
        const dataToFormData = new FormData();
        if (image !== null) {
            dataToFormData.append("image", image, image.name);
        }
        if (tags !== undefined || tags.length !== 0) {
            tags.map(tag => {
                dataToFormData.append("tag_list", tag.text);
                return tag.text;
            });
        }
        dataToFormData.append("title", title);
        dataToFormData.append("body", body);
        dataToFormData.append("description", description);

        this.props.onCreateArticle(dataToFormData);
    }

    handleDelete = i => {
        const { tags } = this.state;
        this.setState({
            tags: tags.filter((tag, index) => index !== i),
        });
    }
    
    handleAddition = tag => {
        this.setState(state => ({ tags: [...state.tags, tag] }));
    }
    
    handleDrag = (tag, currPos, newPos) => {
        const { tags } = this.state;
        const stateTags = [...tags];
        const newTags = stateTags.slice();
    
        newTags.splice(currPos, 1);
        newTags.splice(newPos, 0, tag);
    
        // re-render
        this.setState({ tags: newTags });
    }

    render() {
        this.isAuthenticated = localStorage.getItem("token") !== null;
        const { loading, message } = this.props;
        const { disabled, editorState, suggestions, tags, articleData } = this.state;
        const { title, description, body } = articleData;

        this.createArticle = (
            <div>
                <CreateArticleForm
                    editorState={editorState}
                    onChange={this.onChangeHandler}
                    inputChanged={this.inputHandler}
                    submitData={this.createArticleSubmitHandler}
                    buttonAction="Create Article"
                    formTitle="Create an article with Authors Haven"
                    title={title}
                    description={description}
                    body={body}
                    loading={loading}
                    disabled={!disabled}
                    bodyChanged={this.bodyChangedHandler}
                    onImageUpload={this.inputHandler}
                    suggestions={suggestions}
                    tags={tags}
                    handleDelete={this.handleDelete}
                    handleAddition={this.handleAddition}
                    handleDrag={this.handleDrag}
                    handleTagClick={this.handleTagClick}
                    createMode
                />
            </div>
        );
        if (
            message && message ===
            "Article created successfully") {
            successToast(message);
            this.createArticle = (<Redirect to="/" />);
        } else if (!this.isAuthenticated) {
            this.createArticle = (<Redirect to="/login" />);
        }

        return (
            <div>
                {this.createArticle}
            </div>
        );
    }
}

export const mapStateToProps = state => {
    return {
        loading: state.article.loading,
        article: state.article.article,
        message: state.article.message
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        onCreateArticle: data => dispatch(actions.createArticle(data))
    };
};

CreateArticle.propTypes = {
    loading: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateArticle);
