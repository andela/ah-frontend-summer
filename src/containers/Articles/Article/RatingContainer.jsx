import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import Rating from '../../../components/Rating';
import updateRateValue  from '../../../store/actions/async/rate';


export class RatingContainer extends Component {
    state = {
        rateScore: 0
    }
    static defaultProps = {
        rateScore: 0,
        match: {},
        rate: {},
        onStarClick: () => { },
        averageRating: 0
    };
    
    onStarClick = (nextValue) => {
        const { 
            match: {params: {slug}} = {params: {}},
            rate} = this.props;
        const url =
            `https://ah-backend-summer-staging.herokuapp.com/api/v1/articles/${slug}/rate`;
        this.setState({ rateScore: nextValue });
        const data = { rate_score: nextValue };
        const payload = {
            data,
            url
        };
        rate(payload);
    }
    render() {
        const { rateScore } = this.state;
        const {
            averageRating
        } = this.props;

        const ratingProps = {
            rateScore,
            onStarClick: this.onStarClick,
            averageRating
        };
        return (
            <div>
                <Rating {...ratingProps} />
            </div>
        );

    }

}
export const mapDispatchToProps = dispatch => {
    return {
        rate: payload => dispatch(updateRateValue(payload))
    };
};

export const mapStateToProps = state => {
    return {
        averageRating: state.article.averageRating
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(RatingContainer);

RatingContainer.propTypes = {
    rate: PropTypes.func,
    rateScore: PropTypes.number,
    onStarClick: PropTypes.func,
    averageRating: PropTypes.number,
    match: PropTypes.object
};
