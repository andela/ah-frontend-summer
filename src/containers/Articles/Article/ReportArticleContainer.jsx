import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { reportArticle } from '../../../store/actions/async/ReportArticle';
import ReportArticle from '../../../components/UI/Article/ReportArticle';

export class ReportArticleContainer extends Component {
    state = {
        reason: ''
    };

    onChangeHandler = (e, { value }) => this.setState({ reason: value });

    onSubmitHandler = e => {
        e.preventDefault();
        const { 
            reportArticle, 
            match: { params: { slug } = {} } = {}, 
            history: { push } = {}
        } = this.props;
        const { reason } = this.state;
        const token = localStorage.getItem('token');
        const data = { reason }; 
        reportArticle({ token, data, slug, push });
    };

    render() {
        const { reason } = this.state;
        const { match: { params: { slug } = {} } = {} } = this.props;
        const reportArticleProps = {
            onSubmitHandler: this.onSubmitHandler,
            onChangeHandler: this.onChangeHandler,
            slug,
            reason
        };
        return <ReportArticle {...reportArticleProps}/>;
    }
}

const mapDispatchToProps = dispatch => {
    return {
        reportArticle: payload => dispatch(reportArticle(payload))
    };
};

ReportArticleContainer.propTypes = {
    reportArticle: PropTypes.func.isRequired,
    match: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired
};

export default connect(null, mapDispatchToProps)(ReportArticleContainer);
