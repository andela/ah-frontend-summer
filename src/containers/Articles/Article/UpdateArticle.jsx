import React, { Component } from "react";
import { connect } from "react-redux";
import { EditorState } from "draft-js";
import { Redirect } from "react-router-dom";

import { successToast } from "../../../components/Toast/toast";
import * as actions from "../../../store/actions/index";
import CreateArticleForm from "../../../components/UI/Article/ArticleForm";

export class UpdateArticle extends Component {

    state = {
        editorState: EditorState.createEmpty(),
        articleData: {
            title: "",
            description: "",
            body: "",
            image: null
        },
        disabled: false
    }

    componentDidMount() {
        const articleData = { ...this.state.articleData };
        if (this.props.article) {
            const { title, description, body, image } = this.props.article;
            articleData.title = title;
            articleData.description = description;
            articleData.body = body;
            articleData.image = image;
            this.setState({
                articleData
            });
        };
    }

    bodyChangedHandler = (event, editor) => {
        const data = editor.getData();
        const articleData = { ...this.state.articleData };
        let disableBtn = true;
        articleData.body = data;
        const { title, description, body } = articleData

        if ( !title || !description || !body) {
            disableBtn = false
        };
        if (!body || body.length < 5) {
            disableBtn = false
        };

        this.setState({
            articleData,
            disabled: disableBtn
        });
    }

    inputHandler = (e) => {
        const articleData = { ...this.state.articleData };
        let disableBtn = true;
        const { value, name, files } = e.target

        if (name === "title") {
            articleData.title = value
        } else if (name === "description") {
            articleData.description = value
        };
        const { body, title, description } = articleData

        if ( !body || body < 5 || !title || !description) {
            disableBtn = false
        };
        if (files) {
            articleData.image = files[0]
        };

        this.setState({
            articleData,
            disabled: disableBtn
        });
        e.preventDefault();
    }

    updateArticleHandler = () => {
        const dataToFormData = new FormData();
        const {
            image,
            title,
            body,
            description
        } = this.state.articleData
        if (image) {
            dataToFormData.append("image", image, image.name);
        };
        dataToFormData.append("title", title);
        dataToFormData.append("body", body);
        dataToFormData.append("description", description);

        this.props.onUpdateArticle(this.props.match.params.slug, dataToFormData);

        if (!this.props.loading) {
            successToast("Changes have been saved successfully");
            this.props.history.push(`/articles/${this.props.match.params.slug}`);
        };
    }

    render() {
        const { title, description, body } = this.state.articleData;

        this.updateForm = (
            <div>
                <CreateArticleForm
                    editorState={this.state.editorState}
                    inputChanged={this.inputHandler}
                    submitData={this.updateArticleHandler}
                    title={title}
                    description={description}
                    body={body}
                    loading={this.props.loading}
                    bodyChanged={this.bodyChangedHandler}
                    formTitle="Update Article"
                    buttonAction="Save Changes"
                    disabled={!this.state.disabled}
                    onImageUpload={this.inputHandler}
                />
            </div>
        );

        if (this.props.article === null) {
            this.updateForm = (<Redirect to="/" />);
        };

        return (
            <div>
                {this.updateForm}
            </div>
        );
    }
}

export const mapStateToProps = (state) => {
    return {
        article: state.article.article,
        loading: state.article.loading
    };
};

export const mapDispatchToProps = (dispatch) => {
    return {
        onUpdateArticle: (slug, data) => dispatch(actions.updateArticle(slug, data))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(UpdateArticle);
