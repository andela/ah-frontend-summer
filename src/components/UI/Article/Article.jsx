import React from 'react';
import { Container, Button, Header, Image, Icon } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import renderHTML from 'react-render-html';

import '../../../assets/styles/Article.scss';

const article = (props) => {
    const {
        title,
        author,
        dateCreated,
        readTime,
        isAuthor,
        deleteArticle,
        slug,
        loading,
        image,
        body,
        facebook,
        twitter,
        email
    } = props;
    const renderedBody = renderHTML(body)
    const defautImage = 'https://images.pexels.com/photos/1591057/pexels-photo-1591057.jpeg?cs=srgb&dl=buttons-characters-concept-1591057.jpg&fm=jpg';
    const isLoggedIn = localStorage.getItem('token');

    return (
        <div>
            <Container>
                <Header as='h1' padded="very">
                    {title}
                </Header>
            </Container>
            <Container>
                {author}
                <Container>
                    {dateCreated}
                    <p>{readTime}</p>
                </Container>
            </Container>
            {isAuthor ? (
                <div className="actionButtons">
                    <Container textAlign='right'>
                        <Button
                            color="red"
                            basic
                            onClick={deleteArticle}>
                            Delete
                        </Button>
                        <Link className="article-card-title" to={`/articles/update/${slug}`}>
                            <Button color="teal" basic loading={loading}>Update</Button>
                        </Link>
                    </Container>
                </div>
            ) : null
            }
            <Container textAlign='justified'>
                <Image src={image ? image : defautImage} />
                <div className="ArticleBody">
                    {renderedBody}
                </div>
            </Container>
            {isLoggedIn ? (
                <Container textAlign="center">
                    <a href={facebook} className="facebook"><Icon name="facebook f" /></a>
                    <a href={twitter} className="twitter"><Icon name="twitter" /></a>
                    <a href={email} className="email"><Icon name="mail" /></a>
                </Container>
            ) : ""}
        </div>
    );
};

export default article;
