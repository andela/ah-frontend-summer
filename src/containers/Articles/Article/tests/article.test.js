import React from 'react';
import { shallow } from "enzyme";

import Article from '../Article';

describe('render Article Container', () => {
    const wrapper = shallow(<Article />);
    it("should render rate component correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });
});
