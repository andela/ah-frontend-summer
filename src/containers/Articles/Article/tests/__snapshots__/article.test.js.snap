// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Article Container should render rate component correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
